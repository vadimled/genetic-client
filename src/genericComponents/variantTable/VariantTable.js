import React, { Component } from "react";
import PropTypes from "prop-types";
import { Table, Tooltip, Checkbox } from "antd";
import { Resizable } from "react-resizable";
import SimpleSelect from "GenericComponents/simpleSelect";
import ConfirmationStatus from "GenericComponents/confirmationStatus";
import Notes from "GenericComponents/notes";
import {
  ZYGOSITY_OPTIONS,
  GERMLINE_VARIANT_CLASS_OPTIONS,
  SOMATIC_VARIANT_CLASS_OPTIONS
} from "Utils/constants";
import ExternalLink from "GenericComponents/externalLink";

const ResizeableTitle = props => {
  const { onResize, width, ...restProps } = props;

  if (!width) {
    return <th {...restProps} />;
  }

  return (
    <Resizable width={width} height={0} onResize={onResize}>
      <th {...restProps} />
    </Resizable>
  );
};

class VariantTable extends Component {
  state = {
    columns: [
      {
        key: "1",
        dataIndex: "selection",
        width: 40,
        fixed: "left",
        className: "selection-cell",
        render: (text, record) => {
          if (record.status) {
            return (
              <ConfirmationStatus
                status={record.status}
                handleStatus={(status) => this.props.handleConfirmationStatus({
                  id: record.id,
                  status
                })}
              />
            );
          }
          return (
            <Checkbox
              checked={record.selected}
              onChange={this.props.handleSelectedRow.bind(null, {
                item: record,
                value: !record.selected
              })}
            />
          );
        }
      },
      {
        title: "Gene",
        dataIndex: "gene",
        key: "2",
        width: 200
      },
      {
        title: "Chr: position",
        dataIndex: "chrPosition",
        key: "3",
        width: 200
      },
      {
        title: "Transcript",
        dataIndex: "transcript",
        key: "4",
        width: 200
      },
      {
        title: "Exon",
        dataIndex: "exon",
        key: "5",
        width: 100
      },
      {
        title: "Allele change",
        dataIndex: "alleleChange",
        key: "6",
        width: 100,

        render: (text, record) => {
          return (
            <Tooltip placement="topLeft" title={record.alleleChangeLong}>
              <div>{text}</div>
            </Tooltip>
          );
        }
      },
      {
        title: "coding",
        dataIndex: "coding",
        key: "7",
        width: 100,
        render: (text, record) => {
          return (
            <Tooltip placement="topLeft" title={record.codingLong}>
              <div>{text}</div>
            </Tooltip>
          );
        }
      },
      {
        title: "Protein",
        dataIndex: "protein",
        key: "8",
        width: 100
      },
      {
        title: "VAF",
        dataIndex: "vaf",
        key: "9",
        width: 100
      },
      {
        title: "Zygosity",
        dataIndex: "zygosity",
        key: "10",
        width: 150
      },
      {
        title: "Variant Class",
        dataIndex: "variantClass",
        key: "11",
        width: 170
      },
      {
        title: "coverage",
        dataIndex: "coverage",
        key: "12",
        width: 100
      },
      {
        title: "Notes",
        dataIndex: "notes",
        key: "13",
        width: 532
      },
      {
        title: "Activity log",
        dataIndex: "activityLog",
        key: "14",
        width: 200
      }
    ]
  };

  components = {
    header: {
      cell: ResizeableTitle
    }
  };

  columnsConverter = columns => {
    return columns.map((col, index) => {
      let column = {
        ...col,
        onHeaderCell: column => {
          return {
            width: column.width,
            onResize: this.handleResize(index)
          };
        }
      };

      if (column.dataIndex === "selection") {
        column.title = <div className="table-header-selection-chbx">
          <Checkbox
            checked={this.props.isAllRowSelected}
            onChange={this.props.handleSelectAllRows.bind(null, this.props.isAllRowSelected)}
          />
        </div>;
      }

      if (column.dataIndex === "zygosity") {
        column.render = (...data) => (
          <div className="table-select-wrapper">
            <SimpleSelect
              value={data[1].zygosity}
              options={ZYGOSITY_OPTIONS}
              onChange={e =>
                this.props.handleZygosity({
                  item: data[1],
                  value: e.target.value
                })
              }
              isClearAvailable
            />
          </div>
        );
        column.className = "select";
      }

      if (column.dataIndex === "variantClass") {
        column.render = (...data) =>
          data[1].zygosity &&
          data[1].zygosity !== "insignificant" &&
          data[1].zygosity !== "notReal" &&
          data[1].zygosity !== "unknown" ? (
              <div className="table-select-wrapper">
                <SimpleSelect
                  value={data[1].variantClass}
                  options={
                    data[1].zygosity === "somatic"
                      ? SOMATIC_VARIANT_CLASS_OPTIONS
                      : GERMLINE_VARIANT_CLASS_OPTIONS
                  }
                  onChange={e =>
                    this.props.handleVariantClass({
                      item: data[1],
                      value: e.target.value
                    })
                  }
                />
              </div>
            ) : (
              ""
            );
        column.className = "select";
      }

      if (col.dataIndex === "notes") {
        column.render = (...data) => <Notes
          setNotes={notes => this.props.setNotes({
            id: data[1].id,
            notes
          })}
          value={data[1].notes}
        />;
      }

      if (col.dataIndex === "transcript") {
        column.render = (...data) => <ExternalLink data={data[1].transcript} />;
      }

      if (col.dataIndex === "chrPosition") {
        column.render = (...data) => {
          const { chrPosition } = data[1];
          return (
            <ExternalLink
              data={chrPosition}
              externalHandler={this.props.handelChrPosition.bind(
                null,
                chrPosition
              )}
            />
          );
        };
      }

      return column;
    });
  };

  handleResize = index => (e, { size }) => {
    this.setState(({ columns }) => {
      const nextColumns = [...columns];
      nextColumns[index] = {
        ...nextColumns[index],
        width: size.width
      };
      return { columns: nextColumns };
    });
  };

  render() {
    const { data } = this.props;

    // add options to columns
    const columns = this.columnsConverter(this.state.columns);

    return (
      <Table
        components={this.components}
        pagination={{ pageSize: 20 }}
        bordered
        columns={columns}
        dataSource={data}
        scroll={{ x: "max-content", y: "true" }}
      />
    );
  }
}

VariantTable.propTypes = {
  data: PropTypes.array,
  handleSelectedRow: PropTypes.func.isRequired,
  handleSelectAllRows: PropTypes.func.isRequired,
  handleZygosity: PropTypes.func.isRequired,
  handleVariantClass: PropTypes.func.isRequired,
  handelChrPosition: PropTypes.func.isRequired,
  handleConfirmationStatus: PropTypes.func.isRequired,
  isAllRowSelected: PropTypes.bool,
  setNotes: PropTypes.func.isRequired,
};

VariantTable.defaultProps = {
  data: [],
  isAllRowSelected: false
};

export default VariantTable;
