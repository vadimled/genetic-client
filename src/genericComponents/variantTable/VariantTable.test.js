import React from "react";
import { render } from "@testing-library/react";
import "jest-dom/extend-expect";
import VariantTable from './VariantTable';

describe('VariantTable', () => {
  it('VariantTable snapshot', () => {
    const { asFragment } = render(
      <VariantTable
        onSelectRowKey={() => true}
        handleZygosity={() => true}
        handleVariantClass={() => true}
        handelChrPosition={() => true}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
