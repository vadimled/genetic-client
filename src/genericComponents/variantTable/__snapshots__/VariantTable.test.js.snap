// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantTable VariantTable snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-table-wrapper"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-default ant-table-bordered ant-table-empty ant-table-fixed-header ant-table-scroll-position-left ant-table-scroll-position-right"
        >
          <div
            class="ant-table-content"
          >
            <div
              class="ant-table-scroll"
            >
              <div
                class="ant-table-header"
              >
                <table
                  class="ant-table-fixed"
                >
                  <colgroup>
                    <col
                      class="ant-table-selection-col"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 150px; min-width: 150px;"
                    />
                    <col
                      style="width: 170px; min-width: 170px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 532px; min-width: 532px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-fixed-columns-in-body ant-table-selection-column"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              <div
                                class="ant-table-selection"
                              >
                                <label
                                  class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
                                >
                                  <span
                                    class="ant-checkbox ant-checkbox-disabled"
                                  >
                                    <input
                                      class="ant-checkbox-input"
                                      disabled=""
                                      type="checkbox"
                                      value=""
                                    />
                                    <span
                                      class="ant-checkbox-inner"
                                    />
                                  </span>
                                </label>
                              </div>
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Gene
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Chr: position
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Transcript
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Exon
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Allele change
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              coding
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Protein
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              VAF
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="select react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Zygosity
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="select react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Variant Class
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              coverage
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Notes
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                      <th
                        class="react-resizable"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              Activity log
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                        <span
                          class="react-resizable-handle react-resizable-handle-se"
                        />
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="ant-table-body"
                style="overflow-x: scroll; max-height: true; overflow-y: scroll;"
              >
                <table
                  class="ant-table-fixed"
                >
                  <colgroup>
                    <col
                      class="ant-table-selection-col"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 150px; min-width: 150px;"
                    />
                    <col
                      style="width: 170px; min-width: 170px;"
                    />
                    <col
                      style="width: 100px; min-width: 100px;"
                    />
                    <col
                      style="width: 532px; min-width: 532px;"
                    />
                    <col
                      style="width: 200px; min-width: 200px;"
                    />
                  </colgroup>
                  <tbody
                    class="ant-table-tbody"
                  />
                </table>
              </div>
              <div
                class="ant-table-placeholder"
              >
                <div
                  class="ant-empty ant-empty-normal"
                >
                  <div
                    class="ant-empty-image"
                  >
                    <img
                      alt="No Data"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNDEiIHZpZXdCb3g9IjAgMCA2NCA0MSIgIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiAgPGcgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoMCAxKSIgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj4KICAgIDxlbGxpcHNlIGZpbGw9IiNGNUY1RjUiIGN4PSIzMiIgY3k9IjMzIiByeD0iMzIiIHJ5PSI3Ii8+CiAgICA8ZyBmaWxsLXJ1bGU9Im5vbnplcm8iIHN0cm9rZT0iI0Q5RDlEOSI+CiAgICAgIDxwYXRoIGQ9Ik01NSAxMi43Nkw0NC44NTQgMS4yNThDNDQuMzY3LjQ3NCA0My42NTYgMCA0Mi45MDcgMEgyMS4wOTNjLS43NDkgMC0xLjQ2LjQ3NC0xLjk0NyAxLjI1N0w5IDEyLjc2MVYyMmg0NnYtOS4yNHoiLz4KICAgICAgPHBhdGggZD0iTTQxLjYxMyAxNS45MzFjMC0xLjYwNS45OTQtMi45MyAyLjIyNy0yLjkzMUg1NXYxOC4xMzdDNTUgMzMuMjYgNTMuNjggMzUgNTIuMDUgMzVoLTQwLjFDMTAuMzIgMzUgOSAzMy4yNTkgOSAzMS4xMzdWMTNoMTEuMTZjMS4yMzMgMCAyLjIyNyAxLjMyMyAyLjIyNyAyLjkyOHYuMDIyYzAgMS42MDUgMS4wMDUgMi45MDEgMi4yMzcgMi45MDFoMTQuNzUyYzEuMjMyIDAgMi4yMzctMS4zMDggMi4yMzctMi45MTN2LS4wMDd6IiBmaWxsPSIjRkFGQUZBIi8+CiAgICA8L2c+CiAgPC9nPgo8L3N2Zz4K"
                    />
                  </div>
                  <p
                    class="ant-empty-description"
                  >
                    No Data
                  </p>
                </div>
              </div>
            </div>
            <div
              class="ant-table-fixed-left"
            >
              <div
                class="ant-table-header"
              >
                <table
                  class="ant-table-fixed"
                >
                  <colgroup>
                    <col
                      class="ant-table-selection-col"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-selection-column"
                      >
                        <span
                          class="ant-table-header-column"
                        >
                          <div>
                            <span
                              class="ant-table-column-title"
                            >
                              <div
                                class="ant-table-selection"
                              >
                                <label
                                  class="ant-checkbox-wrapper ant-checkbox-wrapper-disabled"
                                >
                                  <span
                                    class="ant-checkbox ant-checkbox-disabled"
                                  >
                                    <input
                                      class="ant-checkbox-input"
                                      disabled=""
                                      type="checkbox"
                                      value=""
                                    />
                                    <span
                                      class="ant-checkbox-inner"
                                    />
                                  </span>
                                </label>
                              </div>
                            </span>
                            <span
                              class="ant-table-column-sorter"
                            />
                          </div>
                        </span>
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="ant-table-body-outer"
              >
                <div
                  class="ant-table-body-inner"
                  style="max-height: true; overflow-y: scroll;"
                >
                  <table
                    class="ant-table-fixed"
                  >
                    <colgroup>
                      <col
                        class="ant-table-selection-col"
                      />
                    </colgroup>
                    <tbody
                      class="ant-table-tbody"
                    />
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
